driver-core/src/test/unit/com/mongodb/internal/connection/ScramShaAuthenticatorSpecification.groovy

```diff
@@ -360,13 +360,44 @@ class ScramShaAuthenticatorSpecification extends Specification {
         async << [true, false]
     }
 
-    def createConnection(List<String> serverResponses) {
+    def 'should complete authentication when done is set to true prematurely SHA-256'() {
+        given:
+        def serverResponses = createMessages('''
+            S: r=rOprNGfwEbeRWgbNEkqO%hvYDpWUa2RaTCAfuxFIlj)hNlF$k0,s=W22ZaJ0SNY7soEsUEjb6gQ==,i=4096
+            S: v=6rriTRBi23WpRR/wtup+mMhUZUn/dB5nLTJRsjl95G4=
+        ''').last()
+        def authenticator = new ScramShaAuthenticator(SHA256_CREDENTIAL, { 'rOprNGfwEbeRWgbNEkqO' }, { 'pencil' })
+
+        when:
+        // server sends done=true on first response, client is not complete after processing response
+        authenticate(createConnection(serverResponses, 0), authenticator, async)
+
+        then:
+        def e = thrown(MongoSecurityException)
+        e.getMessage().contains('server completed challenges before client completed responses')
+
+        when:
+        // server sends done=true on second response, client is complete after processing response
+        authenticate(createConnection(serverResponses, 1), authenticator, async)
```

Comment:
This is the test where the first check of `!saslClient.isComplete()=true` and the check after evaluating the challenge is also true. After processing the server response, the saslClient should be complete.

Reply:
I don't think this proves anything, since the assertion would pass even if evaluateChallenge wasn't called the final time.  To test properly, create a scenario where `ScramShaAuthenticator.ScramShaSaslClient#validateServerSignature actually fails to validate the server signature and throws a `SaslException`.

Reply:
Probably best then to split this into a separate test.

Reply:
I have added a test where evaluateChallenge fails the final time because of an invalid server response.

Reply:
Acknowledged